keps/sig-storage/2485-read-write-once-pod-pv-access-mode/README.md

```diff
@@ -0,0 +1,972 @@
+- Outline changes needed in CSI spec and sidecars to support the
+  ReadWriteOncePod access mode
+- Outline changes needed in CSI spec and sidecars to continue supporting the
+  ReadWriteOnce access mode
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### Kubernetes Changes, Access Mode
+
+In Kubernetes, we should add a new ReadWriteOncePod persistent volume access
+mode to PersistentVolumes and PersistentVolumeClaims. This change will require
+adding a feature gate and updating validation logic to accept this new access
+mode type if the feature gate is enabled.
+
+```golang
+       // can be mounted read/write mode to exactly 1 pod
+       ReadWriteOncePod PersistentVolumeAccessMode = "ReadWriteOncePod"
+```
+
+This access mode will be enforced in two places:
+
+- First is at the time a volume is mounted for filesystem devices, and at the
+  time a volume is mapped for block devices. During the mount operation, kubelet
+  will check the actual state of the world to determine if the volume is already
+  in-use by another pod. If it is, kubelet will fail mounting with an
+  appropriate error message.
+- As an additional precaution, this will also be enforced during scheduling.
+  When scheduling a pod, if another pod is found using the same PVC and the PVC
+  uses ReadWriteOncePod, then scheduling will fail and the pod will be
+  considered unresolvable.
+
+### CSI Specification Changes, Volume Capabilities
+
+In the CSI spec we should add two new access modes that explicitly state the
+number of writers on a single node.
+
+```protobuf
+      // Can only be published once as read/write at a single worklad on
+      // a single node, at any given time.
+      SINGLE_NODE_SINGLE_WRITER = 6;
+
+      // Can be published as read/write at multiple workloads on a
+      // single node simultaneously.
+      SINGLE_NODE_MULTI_WRITER = 7;
+```
+
+These access modes are modeled after the existing `MULTI_NODE_SINGLE_WRITER` and
+`MULTI_NODE_MULTI_WRITER` access modes. The reason for making this distinction
+is because the `SINGLE_NODE_WRITER` volume capability has conflicting
+definitions (see the [motivation](#motivation) section for context).
+
+For CSI clients, the new ReadWriteOncePod Kubernetes access mode will map to the
+`SINGLE_NODE_SINGLE_WRITER` volume capability access mode in the CSI spec.
+
+For the ReadWriteOnce access mode, the value it maps to depends on the CSI
+driver. If the CSI driver supports the `SINGLE_NODE_SINGLE_WRITER` access mode,
```

Comment:
I think it should use `SINGLE_NODE_MULTI_WRITER`, at least it's the current behavior.

Reply:
Yes you're right, this is the wrong mapping. Fixed.